Allow object keys to be verified by schema

rgoble4 opened the pull request:
I saw this issue: https://github.com/hapijs/joi/issues/1382 and thought I might take a shot at it. It adds dynamicKeys to support validating object keys with schema instead of requiring a regex parameter. It also reworks pattern to just use dynamicKeys under the hood, but that might not be a great thing to stick with if there is a large performance gain in testing a single regex vs a schema that has a single regex rule.

~~As part of this, I also updated describe, but cannot for the life of me figure out how to make the describe test pass. Based on the output, the actual and expected describe results are identical, but the test still fails.~~ edit: I just had to sleep on it, found and fixed.

Anyway, I hope this helps or at least gives some folks ideas on how they might want to make this feature happen. If this is even reasonably close to mergeable, I'm happy to keep working on it.

Marsup commented:
I'd like to stay closer to what currently exists, integrating schemas into the current code should be fairly easy, no need for a new API or casting values, I'd just make `pattern()` accept schemas as 1st argument.

rgoble4 commented:
Ah, I got it in my mind that pattern = regex, but it also makes sense that a schema would essentially still be a pattern for the keys. To be clear, should pattern accept standard regex as it does currently as well as a schema in the first arg to preserve backwards compatibility?

rgoble4 commented:
Ok, updated as discussed. I also rebased and started clean so that there weren't a bunch of line diffs that were related to the direction I had previous gone in. Let me know what else you'd like to see.

rgoble4 commented:
Good points, updated!
